### ISSUE TITLE
RFE: optional one-shot/cached macro expansion 

### ISSUE THREAD

**Issue by pmatilai** [rpm-software-management/rpm#1155]
Some macros are really expensive to expand, notorious examples include python/perl/etc related macros that invoke the interpreter to fish out a configuration path or such. When these are expanded repeatedly, it really adds up.

In a spec one can use %global to achieve this, but that's not possible from a macro file which is how most such macros are distributed.

**Comment by voxik** [comment]
Is there any benefit in having macro files? Wouldn't be better to deprecate them instead?

**Comment by pmatilai** [comment]
Um. And use what instead? The spec file format? Whose parser is the one of the most compilicated and loathed piece of code in rpm and would be difficult to lift out of librpmbuild to put it mildly, and slow every rpm startup in the process?

In contrast, the macro file format is a nice, declarative format that can be always safely read without side-effects. It could be extended without sacrificing that.

**Comment by voxik** [comment]
First seeing spec file and then trying to move some macros into macros file, it was very confusing, I can tell you. From packager POV, the macro file is the same as spec file, but you don't write `%global` everywhere for some unknown reason. Also the difference between `%import` and `%load` (where the later was introduced on my request) is not really obvious.

Obviously, you know the internals, but honestly, that is just implementation detail to me.

**Comment by pmatilai** [comment]
Well yes, people occasionally get confused about that. The difference is a big one though, its like comparing a shell script with an ini file. Both have their places.

**Comment by mlschroe** [comment]
Crazy thought: we could allow to use %define/%global in the macro file, as those are illegal macro names. I.e. we could allow this:
```
%foo hello
%define bar world
```
(rpm internally somewhat rewrites the %foo to %define foo anyway)

Then we could also allow
```
%undefine foo
```
which is something I wanted to have since quite some time.

The downside is of course that people will start to try to use %if statements, because the macro file looks too much like a spec file.

**Comment by pmatilai** [comment]
There are other downsides too, such as arbitrary side-effects and code execution on load (as opposed to time of use).

**Comment by mlschroe** [comment]
Well, we don't need to allow %global.

Anyway, let's move a step back and talk about this issue. This is about one shot macros. Implementation wise this is not hard, it's more a question of syntax.

It would be nice to have the following options:
- literal: do not expand the macro when using it
- global: put macro in global namespace
- expand: expand macro right away
- once: cache expanded macro (this issue)

So %global is %define plus the global + expand flags.

Macro names can't start with a `-`, so we could make use normal options:
```
%global foo bar
%define -g -x foo bar
```

As next step can can discuss how we support this in the macro files ;)

**Comment by pmatilai** [comment]
Yeah, %define with options is what I'd really like (I think I even mentioned somewhere surrounding the pending %literal). 

In macro files the name limitation could be abused for similar effect by adding special characters in the front of the name. AIUI rpm5 used "." to denote a read-only macro, for our purposes eg "=" could mean "literal" and so on. It's pretty ugly of course.

I suppose we could place the macro flags after the name, using a separator such as ':'
That would fit in the macro file syntax easily, eg:
```
%_pyver:o %(python3 --version)
```

Not sure its any prettier than special characters in front, though :joy: For a moment I thought about putting the flags inside [] after the name, but I think we want to reserve that syntax for arrays if we ever get around to implement them.

**Comment by pmatilai** [comment]
FWIW, a simple and generic way to implement this on macro level is to redefine the macro as a global on first evaluation, eg:

> %python3_sitearch %{global python3_sitearch %(RPM_BUILD_ROOT= %{__python3} -Esc "import sysconfig; print(sysconfig.get_path('platlib', vars={'platbase': '%{_prefix}', 'base': '%{_prefix}'}))")}%{python3_sitearch}

Optimally what would use %literal (or the equivalent) instead of %global but we don't have that yet...

**Comment by DemiMarie** [comment]
What about rewriting the macro in Lua and stashing the result in a Lua global variable?

**Comment by pmatilai** [comment]
It's possible in Lua of course, the above was to point out a way to do it *without*, because not everybody is content with Lua.

**Comment by pmatilai** [comment]
Discussed the syntax with @dmnks today and came up with a solution that fits nicely into the macro file syntax, and since it works there it can be also used in %define/%global without any further code: add an optional `<modifiers>` field to the syntax, eg:

`%_pyver<ol> %(python3 --version)`

It combines with parametrics too:
`%mymacro(ab:f)<c> ....`

We could also support long names (perhaps later on), eg
`%mymacro<oneshot,const> ...`

`const` is a work-in-progress name for #2314 which suffers from the same issue: the implementation is easy enough, the syntax has been the blocker. And we already have an implementation for `l`iteral, it's just the define syntax that has been missing.

`<>` is used for the modifier because `[]` we might one day want to use that for arrays, and `{}` is already used in the macro body, so this makes these stand out. In the long name version, `,` is the separator to avoid confusion with parametric arguments using `:` and space is what separates the name part from the body, so spaces seem unappropriate.

**Comment by pmatilai** [comment]
So combining the newer needs with the older stuff in this topic, we have something like:
- `c`onst: not overridable (issues #2314 and #4091)
- `e`xpand: expand macro right away
- `g`lobal: put macro in global namespace
- `l`iteral: do not expand the macro when using it
- `o`nce: cache expanded macro (this issue)

But "expand" can't be valid in macro files (reading a macro file must not execute macros), and "global" would seem kinda wrong as a macro modifier, it's a scope thing, it doesn't make sense in a macro file and elsewhere you could already use %global. So I'm going to leave those two out for now at least. Probably these should be %define arguments instead, eg `%define -g -x foo bar` from the earlier discussion.

**Comment by mlschroe** [comment]
Don't forget:
- `u`ndefine: undefine a macro

Btw, at some point in time we thought about using `%mymacro:c(ab:f)`. But I don't mind `<>`.

**Comment by pmatilai** [comment]
Yeah we rediscovered the idea about options after `:`, but `<>` seems like a clearer separation for a human reader if nothing else.

FWIW, I don't think "undefine" belongs here: that's an action, not a macro property.

**Comment by pmatilai** [comment]
I have a working implementation of this that works by pushing the expansion on top of the original one-shot macro (so it's possible to force re-expand by undefining), as a literal macro as that's what I think most users would expect.

I'm just wondering if that should be left for the macro creator afterall, ie `mymacro<o>` would actually keep expanding on each use like the result of a `%global` does, and `mymacro<ol>` would really only expand once. The ability to differentiate seems kinda handy, but then I smell potential for lot of confusion as well.

There are all these related question surrounding `%global` too - does a one-shot macro actually expand twice if defined with `%global`? It seems a bit nuts, but then, one-shot is about the *time of use*, whereas `%global` (and the eventual `%define -e/-x`) are about *time of definition*. 🤔

**Comment by mlschroe** [comment]
So `mymacro<o>` would expand on first use, cache the result, and immediately expand again? I find that a bit weird. I prefer an implicit `<l>`.

Regarding %global/-e and <o>: I think it should expand at definition time and mark the result as oneshot so that it caches the second expansion at macro use time.

**Comment by pmatilai** [comment]
> So mymacro<o> would expand on first use, cache the result, and immediately expand again? I find that a bit weird. I prefer an implicit <l>.

Yeah, the *result* would get expanded. Which I too find weird. That's what happens with %global, but people find that weird as well. So I really prefer the implicit `<l>` as well. 

> Regarding %global/-e and : I think it should expand at definition time and mark the result as oneshot so that it caches the second expansion at macro use time.

Ack, that was my thinking as well.

**Comment by dmnks** [comment]
> I'm just wondering if that should be left for the macro creator afterall, ie `mymacro<o>` would actually keep expanding on each use like the result of a `%global` does, and `mymacro<ol>` would really only expand once. The ability to differentiate seems kinda handy, but then I smell potential for lot of confusion as well.

This almost calls for a slightly different (more granular) set of modifiers:

- `mymacro<z>`: expands on use (deferred, or *lazy* expansion), i.e. same as `%define`
- `mymacro<m>`: expands on definition and caches (*memoizes*) the result, i.e. same as `%global`
- `mymacro<zm>`: lazy, memoized expansion, i.e. expands on the first use and caches the result
- `mymacro<zml>`: lazy, memoized expansion to a literal, i.e. expands on the first use and caches the result as a literal macro (the original one-shot proposal)

The one-shot modifier (`<o>`) would then just be an alias for `<zml>`.

**Comment by pmatilai** [comment]
We can't allow expand-on-define in the modifier syntax because reading macro files must not execute code.
While it would be possible to make the literalness of the one-shot expansion optional, the *default* needs to be literal because that's the only sane thing, see  the last few comments. The %global behavior is a weird historical artifact that continues to confuse packagers to this day, new things should do what makes most sense to their users.

**Comment by pmatilai** [comment]
I'm more tempted to disable one-shot in combination with %global entirely, which would make that particular dark corner go away.

**Comment by pmatilai** [comment]
Once we have the %global behavior available as %define options (#4163), the behavioral combinations become saner. I have a patch for that too, but it's too closely related to wrestle at the same time as these other macro PR's, and I've too many PR's in flight as it is.

**Comment by pmatilai** [comment]
Basically: there's nothing wrong with declaring a one-shot macro in the global scope. That's where they normally are, the only exception is inside parametric macros. It's the one-shot macro getting double-expanded with %global that's just ... weird. It becomes rather apparent when trying to explain it in the docs 🤪

**Comment by dmnks** [comment]
> We can't allow expand-on-define in the modifier syntax because reading macro files must not execute code.

Oh indeed, good point. I forgot about that...

> While it would be possible to make the literalness of the one-shot expansion optional, the _default_ needs to be literal because that's the only sane thing, see the last few comments.

Yep, my suggestion was just about making it possible, not the default 😄 

> The %global behavior is a weird historical artifact that continues to confuse packagers to this day, new things should do what makes most sense to their users.

True, no point in reaffirming that use case then 😅

**Comment by dmnks** [comment]
> Basically: there's nothing wrong with declaring a one-shot macro in the global scope. That's where they normally are, the only exception is inside parametric macros. It's the one-shot macro getting double-expanded with %global that's just ... weird. It becomes rather apparent when trying to explain it in the docs 🤪

Indeed, "one-shot" somehow makes you think of the (peculiar) `%global` expansion semantics while being a separate thing, which is confusing for sure.

**Comment by pmatilai** [comment]
Yup. Most people don't realize what you defined with %global also keeps getting re-expanded on each use. It's such a weird thing.
